File: pml/sprites/components/mouth.py
"""Mouth component — generates mouth GraphicObjects in various styles."""

from __future__ import annotations

from typing import Any

from pml.graphics.objects import GraphicObject
from pml.sprites.validator import ParamSchema, validate_params

from .view_utils import VIEW_NAMES, sym_str

_MOUTH_SCHEMA = (
    ParamSchema()
    .enum("style", ["neutral", "smile", "frown", "open", "cat", "fang"], "neutral")
    .number("size", 1.0, min_val=0.3, max_val=3.0)
    .enum("view", VIEW_NAMES, "front")
)


def create_mouth(**kwargs: Any) -> GraphicObject:
    """Create a mouth graphic object.

    Args:
        :style — 'neutral | 'smile | 'frown | 'open | 'cat | 'fang
        :size — scale factor (default 1.0)
        :view — 'front | 'side | 'back | 'three-quarter
    """
    p = validate_params(_MOUTH_SCHEMA, {sym_str(k): v for k, v in kwargs.items()})

    style = p["style"]
    s = p["size"]
    view = p["view"]

    if view == "back":
        return GraphicObject(
            shape_type="group",
            children=(),
            metadata={"component": "mouth", "style": style, "view": view},
        )

    if view == "side":
        w = 8 * s
        return GraphicObject(
            shape_type="group",
            children=[
                GraphicObject(
                    shape_type="line",
                    params={"x1": w * 0.2, "y1": 0, "x2": w, "y2": -2 * s},
                    stroke="#c0392b",
                    stroke_width=2.0 * s,
                ),
            ],
            metadata={"component": "mouth", "style": style, "view": view},
        )

    if view == "three-quarter":
        s = s * 0.85  # slightly smaller

    # Front view (and three-quarter): original style logic unchanged
    w = 12 * s
    children: list[GraphicObject] = []

    if style == "neutral":
        children.append(GraphicObject(
            shape_type="line",
            params={"x1": -w / 2, "y1": 0, "x2": w / 2, "y2": 0},
            stroke="#c0392b",
            stroke_width=2.0 * s,
        ))

    elif style == "smile":
        children.append(GraphicObject(
            shape_type="line",
            params={"x1": -w / 2, "y1": -2 * s, "x2": 0, "y2": 4 * s},
            stroke="#c0392b",
            stroke_width=2.0 * s,
        ))
        children.append(GraphicObject(
            shape_type="line",
            params={"x1": 0, "y1": 4 * s, "x2": w / 2, "y2": -2 * s},
            stroke="#c0392b",
            stroke_width=2.0 * s,
        ))

    elif style == "frown":
        children.append(GraphicObject(
            shape_type="line",
            params={"x1": -w / 2, "y1": 4 * s, "x2": 0, "y2": -2 * s},
            stroke="#c0392b",
            stroke_width=2.0 * s,
        ))
        children.append(GraphicObject(
            shape_type="line",
            params={"x1": 0, "y1": -2 * s, "x2": w / 2, "y2": 4 * s},
            stroke="#c0392b",
            stroke_width=2.0 * s,
        ))

    elif style == "open":
        children.append(GraphicObject(
            shape_type="ellipse",
            params={"cx": 0, "cy": 2 * s, "rx": w / 2, "ry": 6 * s},
            fill="#3d1010",
            stroke="#c0392b",
            stroke_width=1.5,
        ))
        children.append(GraphicObject(
            shape_type="rect",
            params={"x": -w / 3, "y": -1 * s, "w": w * 2 / 3, "h": 3 * s},
            fill="#FFFFFF",
        ))

    elif style == "cat":
        cat_w = 8 * s
        for dx in (-cat_w, 0, cat_w):
            children.append(GraphicObject(
                shape_type="ellipse",
                params={"cx": dx, "cy": 2 * s, "rx": 4 * s, "ry": 3 * s},
                fill=None,
                stroke="#c0392b",
                stroke_width=1.5,
            ))

    elif style == "fang":
        children.append(GraphicObject(
            shape_type="line",
            params={"x1": -w / 2, "y1": -2 * s, "x2": w / 2, "y2": -2 * s},
            stroke="#c0392b",
            stroke_width=2.0 * s,
        ))
        fang_x = w / 3
        children.append(GraphicObject(
            shape_type="polygon",
            params={"points": [
                (fang_x - 3 * s, -2 * s),
                (fang_x + 3 * s, -2 * s),
                (fang_x, 6 * s),
            ]},
            fill="#FFFFFF",
            stroke="#c0392b",
            stroke_width=1.0,
        ))

    return GraphicObject(
        shape_type="group",
        children=tuple(children),
        metadata={"component": "mouth", "style": style, "view": view},
    )

File: pml-cpp/src/pml/sprites/components/mouth.cpp
// ═══════════════════════════════════════════════════════════════════════════════
// PML Mouth Component — Implementation
// ═══════════════════════════════════════════════════════════════════════════════

#include "mouth.h"

#include <vector>

namespace pml {

// ═══════════════════════════════════════════════════════════════════════════════
// Schema
// ═══════════════════════════════════════════════════════════════════════════════

ParamSchema mouth_schema() {
    return ParamSchema()
        .enum_("style", {"neutral", "smile", "frown", "open", "cat", "fang"}, "neutral")
        .number("size", 1.0, 0.3, 3.0);
}

// ═══════════════════════════════════════════════════════════════════════════════
// Factory
// ═══════════════════════════════════════════════════════════════════════════════

std::shared_ptr<GraphicObject> create_mouth(
    const std::unordered_map<std::string, Value>& kwargs) {
    auto p = validate_params(mouth_schema(), kwargs);

    std::string style = std::get<std::string>(p["style"]);
    double s = std::get<double>(p["size"]);
    double w = 12.0 * s;

    std::vector<GraphicObject> children;

    if (style == "neutral") {
        children.emplace_back(
            "line",
            std::unordered_map<std::string, Value>{{"x1", -w / 2.0}, {"y1", 0.0}, {"x2", w / 2.0}, {"y2", 0.0}},
            std::nullopt,
            "#c0392b",
            2.0 * s);
    } else if (style == "smile") {
        children.emplace_back(
            "line",
            std::unordered_map<std::string, Value>{{"x1", -w / 2.0}, {"y1", -2.0 * s}, {"x2", 0.0}, {"y2", 4.0 * s}},
            std::nullopt,
            "#c0392b",
            2.0 * s);
        children.emplace_back(
            "line",
            std::unordered_map<std::string, Value>{{"x1", 0.0}, {"y1", 4.0 * s}, {"x2", w / 2.0}, {"y2", -2.0 * s}},
            std::nullopt,
            "#c0392b",
            2.0 * s);
    } else if (style == "frown") {
        children.emplace_back(
            "line",
            std::unordered_map<std::string, Value>{{"x1", -w / 2.0}, {"y1", 4.0 * s}, {"x2", 0.0}, {"y2", -2.0 * s}},
            std::nullopt,
            "#c0392b",
            2.0 * s);
        children.emplace_back(
            "line",
            std::unordered_map<std::string, Value>{{"x1", 0.0}, {"y1", -2.0 * s}, {"x2", w / 2.0}, {"y2", 4.0 * s}},
            std::nullopt,
            "#c0392b",
            2.0 * s);
    } else if (style == "open") {
        children.emplace_back(
            "ellipse",
            std::unordered_map<std::string, Value>{{"cx", 0.0}, {"cy", 2.0 * s}, {"rx", w / 2.0}, {"ry", 6.0 * s}},
            "#3d1010",
            "#c0392b",
            1.5);
        children.emplace_back(
            "rect",
            std::unordered_map<std::string, Value>{{"x", -w / 3.0}, {"y", -1.0 * s}, {"w", w * 2.0 / 3.0}, {"h", 3.0 * s}},
            "#FFFFFF");
    } else if (style == "cat") {
        double cat_w = 8.0 * s;
        for (double dx : {-cat_w, 0.0, cat_w}) {
            children.emplace_back(
                "ellipse",
                std::unordered_map<std::string, Value>{{"cx", dx}, {"cy", 2.0 * s}, {"rx", 4.0 * s}, {"ry", 3.0 * s}},
                std::nullopt,
                "#c0392b",
                1.5);
        }
    } else if (style == "fang") {
        children.emplace_back(
            "line",
            std::unordered_map<std::string, Value>{{"x1", -w / 2.0}, {"y1", -2.0 * s}, {"x2", w / 2.0}, {"y2", -2.0 * s}},
            std::nullopt,
            "#c0392b",
            2.0 * s);
        double fang_x = w / 3.0;
        children.emplace_back(
            "polygon",
            std::unordered_map<std::string, Value>{
                {"points",
                 make_list_value(std::vector<Value>{
                     fang_x - 3.0 * s, -2.0 * s,
                     fang_x + 3.0 * s, -2.0 * s,
                     fang_x, 6.0 * s})}},
            "#FFFFFF",
            "#c0392b",
            1.0);
    }

    return std::make_shared<GraphicObject>(
        "group",
        std::unordered_map<std::string, Value>{},
        std::nullopt,
        std::nullopt,
        1.0,
        AffineTransform::identity(),
        std::move(children),
        std::unordered_map<std::string, Value>{
            {"component", std::string("mouth")},
            {"style", style}});
}

}  // namespace pml
